Exclude prefixes or folders from versioning (#769)
minio/docs: source/administration/object-management/object-versioning.rst

Objects created prior to enabling versioning have a 
``null`` :ref:`version ID <minio-bucket-versioning-id>`.

Exclude a Prefix From Versioning
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can exclude certain :ref:`prefixes <minio-admin-concepts-organize-objects>` from versioning using the :ref:`MinIO Client <minio-client>`.
This is useful for Spark/Hadoop workloads or others that initially create objects with temporary prefixes. 

.. admonition:: Object locking
   :class: note

   Buckets with :ref:`object locking enabled <minio-object-locking>` require versioning and do not support excluding prefixes.

- Use :mc-cmd:`mc version enable` with the :mc-cmd:`~mc version --excluded-prefixes` option:

  .. code-block:: shell
     :class: copyable

>      mc version enable --excluded-prefixes ALIAS/BUCKET "prefix1, prefix2"

Reviewer: I think the syntax has to be:

```suggestion
mc version enable --excluded-prefixes "prefix1,prefix2" ALIAS/BUCKET
```

```
mc version enable --excluded-prefixes play/version-test "foo"                             1 ✘  docs   3.8.16   
mc: <ERROR> Unable to enable versioning: `SetVersion` is not supported for `filesystem`.
```

I _think_ we generally recommend doing all params prior to the ALIAS/BUCKET for parsing reasons, but I am not entirely sure if that's stlil a requirement.

  - Replace ``ALIAS`` with the :mc:`alias <mc alias>` of a configured MinIO deployment.

>   - Replace ``BUCKET`` with the :mc-cmd:`target bucket <mc version ALIAS>` you want to exclude :ref:`prefixes <minio-admin-concepts-organize-objects>` for.

Author: I tried to both follow the model of the existing example above, and link to something useful about prefixes/folders. Ideas? This seems a bit awkward.

Reviewer 2: Couple of options here.

## Link to the S3 Docs
`Bucket` is a standard s3 concept. We could link to their docs explanation.
If you look in conf.py under extlinks, we have a role for that:
```
:s3-docs:`UsingBucket.html`
```

Basically, s3-docs adds the base URL up through `/latest/userguide/` then attaches whatever we have for the rest of the URL.

## Link to the list command

We could link to the mc list command:
```
:mc:`mc ls`
```

(That would list all buckets and objects at the root of the alias.)

## Link to the admin/concept.rst page

There's a section on that page about organizing objects within folders in buckets. Not ideal, but closest we have there. You'd need to add a directive to that page."

## My preference
I think I prefer linking to the S3 docs, personally.

> The list of :mc-cmd:`~mc version --excluded-prefixes` prefixes match all objects containing the specified strings in their prefix or name, similar to a regular expression of the form ``*prefix*``.

Reviewer 3: In this example, the given excluded prefixes `prefix1`, `prefix2` would match objects similar to the regular expression of the form `prefix*`.
```suggestion
The list of :mc-cmd:`~mc version --excluded-prefixes` prefixes match all objects containing the specified strings in their prefix or name, similar to a regular expression of the form ``prefix*``.
```

To match objects by prefix only, use ``prefix/*``.

For example, the following command excludes any objects containing ``abc`` or ``d3f`` in their prefix or name from versioning:

  .. code-block:: shell
     :class: copyable

>      mc version enable local/test-bucket --excluded-prefixes "abc, d3f"

Author: Is there an established pattern for names/tokens in examples? (Rather than abc, etc.)

Reviewer 2: `prefix1, prefix2`
or
`myprefix, mysecondprefix`

Something that is human readable.

We typically use the word for whatever it is we are samplig. `mybucket`, `myminio` (for an alias), and so on.

You can exclude up to 10 prefixes for each bucket.
To add or remove prefixes, repeat the :mc-cmd:`mc version enable` command with an updated list.
The new list of prefixes replaces the previous one.

> To view the currently excluded prefixes, use :mc-cmd:`mc version enable` with the :mc-cmd:`~mc version enable --JSON` option:

Reviewer 3: ```suggestion
To view the currently excluded prefixes, use :mc-cmd:`mc version info` with the :mc-cmd:`~mc version enable --JSON` option:
```

  .. code-block:: shell
     :class: copyable

     mc version enable --excluded-prefixes ALIAS/BUCKET --json

The command output resembles the following, with the list of excluded prefixes in the ``ExcludedPrefixes`` property:

.. code-block:: shell

     $ mc version info local/test-bucket --json
     {
      "Op": "info",
      "status": "success",
      "url": "local/test-bucket",
      "versioning": {
       "status": "Enabled",
       "MFADelete": "",
       "ExcludedPrefixes": [
        "abc, d3f"
       ]
      }
     }

To disable prefix exclusion and resume versioning all prefixes, repeat the :mc-cmd:`mc version enable` command without :mc-cmd:`~mc version --excluded-prefixes`:

  .. code-block:: shell
     :class: copyable

     mc version enable ALIAS/BUCKET

     
Exclude Folders from Versioning
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can exclude folders from versioning using the :ref:`MinIO Client <minio-client>`.

.. admonition:: Object locking
   :class: note

   Buckets with :ref:`object locking enabled <minio-object-locking>` require versioning and do not support excluding folders.

> - Use :mc-cmd:`mc version enable` with the :mc-cmd:`~mc version --excluded-prefixes` option to exclude objects with names ending in ``/`` from versioning:

Reviewer: We can use a more 'real world' example here.

e.g.

The following :mc-cmd:`mc version enable` command includes the :mc-cmd:`~mc verison --excluded-prefixes` option to exclude objects in the ``_temp`` prefix from versioning:

```
mc version enable --exclude-folders "_temp" myminio/mybucket
```

Or something along those lines. We get to be creative with examples, so feel free to have a bit of fun thinking of an interesting scenario to build the example around.

> - Use :mc-cmd:`mc version enable` with the :mc-cmd:`~mc version --excluded-prefixes` option to exclude objects with names ending in ``/`` from versioning:

Reviewer 3: ```suggestion
- Use :mc-cmd:`mc version enable` with the :mc-cmd:`~mc version --exclude-folders` option to exclude objects with names ending in ``/`` from versioning:
```

Author: @krisis Oooh missed that paste failure thank you!!

  .. code-block:: shell
     :class: copyable

     mc version enable --exclude-folders ALIAS/BUCKET

  - Replace ``ALIAS`` with the :mc:`alias <mc alias>` of a configured MinIO deployment.

  - Replace ``BUCKET`` with the :mc-cmd:`target bucket <mc version ALIAS>` you want to exclude :ref:`folders <minio-admin-concepts-organize-objects>` for.

To check whether folders are versioned for a bucket, use the :mc-cmd:`mc version enable` command with the ``--json`` option.
If the ``ExcludeFolders`` property is ``true``, folders in that bucket are not versioned.

  .. code-block:: shell
     :class: copyable

     mc version enable --excluded-prefixes ALIAS/BUCKET --json

The command output resembles the following:

.. code-block:: shell

     $ mc version info local/test-bucket --json
     {
      "Op": "info",
      "status": "success",

>       "url": "local/test-bucket",

Author: Is there a standard example bucket name I might use here?

Reviewer 2: Most often, we just go with `my-bucket`.

Reviewer: my-bucket, data, etc.

Objects created prior to enabling versioning have a 
``null`` :ref:`version ID <minio-bucket-versioning-id>`.

Exclude a Prefix From Versioning
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can exclude certain prefixes from versioning using the MinIO :mc:`mc` CLI.
This is useful for Spark/Hadoop workloads or others that initially create objects
with temporary prefixes.

- Use the :mc-cmd:`mc version enable` command with the ``--excluded-prefixes``
  option:

  .. code-block:: shell
     :class: copyable

     mc version enable --excluded-prefixes ALIAS/BUCKET "prefix1, prefix2"

  - Replace ``ALIAS`` with the :mc:`alias <mc alias>` of a configured 
    MinIO deployment.

  - Replace ``BUCKET`` with the 
    :mc-cmd:`target bucket <mc version ALIAS>` for which the prefixes should

>     be excluded from versioning.

Author: The existing text just above has the same 
```
:mc-cmd:`target bucket <mc version ALIAS>`
```
maybe that should change too

Something about wildcards and globbing
    
You can exclude up to 10 prefixes for each bucket. To add or remove excluded prefixes,
repeat your ``mc version enable`` command with an updated list. The new list of
prefixes replaces the previous one.

To view the currently excluded prefixes, use the :mc-cmd:`mc version enable`

> command with the ``--json`` option. The ``ExcludedPrefixes`` property contains a list

Reviewer 2: You can simplify this, I think to
> To view the excluded prefixes, use :mc-cmd:`mc version enable` with the  :mc-cmd:`~mc  version enable --json` option.

When we use `~` at the start of the :mc-cmd:, it only displays the last element of the command, so we can link directly to the flag in the docs.

Reviewer 2: Except this doesn't work, because --json is a global flag. So ignore this.

of excluded prefixes:

  .. code-block:: shell
     :class: copyable

     mc version enable --excluded-prefixes ALIAS/BUCKET --json

  .. code-block:: shell

> 

Reviewer 2: Add a sentence at line 306 along the lines of
> The output resembles the following:

     $ mc version info local/test-bucket --json
     {
      "Op": "info",
      "status": "success",
      "url": "local/test-bucket",
      "versioning": {
       "status": "Enabled",
       "MFADelete": "",
       "ExcludedPrefixes": [
        "TempPrefix-"
       ]
      }
     }

Exclude Folders from Versioning
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> You can exclude folders from versioning using the MinIO :mc:`mc` CLI.

Reviewer 2: Only, perhaps also spell `versioning` correctly. 🙄

> - Use the :mc-cmd:`mc version enable` command with the ``--exclude-folders``

Reviewer 2: As above, you can provide a link to the `--exclude-folder` option in the docs:
```
:mc-cmd:`~mc version enable --exclude-folders`
```

And I would drop "the" and "command" and just let the command stand on its own.
> Use `mc version enable` with the `--exclude-folders`...

  option to exclude objects with names ending in `/` from versioning:

  .. code-block:: shell
     :class: copyable

     mc version enable --exclude-folders ALIAS/BUCKET

  - Replace ``ALIAS`` with the :mc:`alias <mc alias>` of a configured
    MinIO deployment.

  - Replace ``BUCKET`` with the

>     :mc-cmd:`target bucket <mc version ALIAS>` for which the folders should

Reviewer 2: Different link for the bucket here, as well.

    be excluded from versioning.

To check whether folders are versioned for a bucket, use the :mc-cmd:`mc version enable`

> command with the ``--json`` option. If the ``ExcludeFolders`` property is ``true``,

Reviewer 2: Similar wording adjustment.

folders in that bucket are not versioned.

  .. code-block:: shell
     :class: copyable

     mc version enable --excluded-prefixes ALIAS/BUCKET --json

>   .. code-block:: shell

Reviewer 2: Add a sentence above intro-ing the example, as in the previous comment above.
